### ISSUE TITLE
Contributions to the GUI

### ISSUE THREAD

**Issue by maximilien-noal** [uxmal/reko#947]
I'd like to contribute to the project of giving Reko a cross-platform GUI (ie. remove Windows Forms), I have a lot of experience in the MVVM pattern and WPF and Windows Forms, and some experience with Blazor.

I already took a look. There's seem to be interfaces everywhere, I assume it's that way to support three use cases (Web driver, CMD driver, WinForms driver), and the GUI-related code seems massive.

First problem  : I don't know where to start.
Second problem : Supporting both AvaloniaUI and Blazor would be as massive as the existing code.
Between those too candidates, was one chosen ?

Supporting both would be kind of a headache. A PWA App with Blazor would simplify things a lot. Also Blazor does not assume MVVM nor MVC. AvaloniaUI is inspired heavily by WPF, and assumes MVVM is used.

Note : if WPF is to be used, it could be converted to a WebApp by this project :
https://github.com/bridgedotnet/Granular
It would mean one GUI, but:
- I did not test it.
- WPF is Windows-centric and "replaced" by UWP (for Windows), .NET MAUI (not released), and WinUI 3 (not released)

**Comment by uxmal** [comment]
Salut Maximilien, great to hear about your interest in the Reko project!

We've been struggling to find a viable cross-platform solution for moving forward the Reko GUI. I've been wanting to move off Windows Forms for a while, but no obvious candidate has appeared. Some of the challenges that a solution must solve is:
1. it must be cross platform. 
1. it must support displaying "rich text", by which I mean text that can be annotated with color, underlines, and superimposed images. An HTML rendering engine would be ideal here.
1. it should support editing limited ranges of rich text. E.g. I'd like to be able to click on a procedure's name in the text view, select a `Rename` menu item, have the text editor indicate that the procedure's name is editable, and as the user types the new name provide validation so that the name is not duplicate, invalid etc.
1. it should support displaying virtual bitmaps. Reko displays "heat maps" and memory maps, which may be very large.
1. it should support the rendering of (directed) graphs. 
1. the GUI framework should be modest in size. We tried Electron a while back, and it consumed several 100's of megabytes just to load, clearly unacceptable.
1. It should still feel like a desktop application. I want menus, keyboard accelerators, Open/Save file capabilities, clipboard access for copy/paste, drag/drop support
1. It should be easy for users to install on a new system. Ideally, there are no manual extra steps for installation.

Right now, @smx-smx and I are experimenting with Chromely, over on https://github.com/uxmal/reko-chromely. It shows some promise, but there is a bunch of low-level hacking to interface between the .NET side and the JavaScript side. As far as I know, Avalonia doesn't have an awesome answer to the rich-text story yet. I haven't sunk my teeth into Blazor yet, but the 'in-process' version of Blazor would also be an awesome alternative -- provided it could be made cross platform.

So, where to start? Well, since you're new to the code base, I can give you a 10,000-meter explanation of how the Reko decompiler could be hosted in any GUI application. It's been designed from the start to not be totally tied to any specific GUI environment. The code decompiler engine communicates with the hosting environment via interfaces. Most of these are located in `src\Core\Services`. They are:
* DecompilerEventListener - the engine uses this interface to notify the user of any errors or warnings, and has a `Canceled` method it polls to stop itself if the user has requested the decompilation to stop.
* IArchiveBrowserService - if a file being opened is in fact an "archive" consiting of multiple files (think disk image, or zip file), this interface displays the contents of the archive for a user to pick a file for decompilation.
* IFileSystemService - all access to storage happens via this interfaces. Normally, the default file system service implementation goes to disk, but you could imagine a "cloud" version where the file system is actually blob storage or even all in plain memory.
* IDecompilerService - this interface is exposed by the Reko decompilation engine to the user interface. It provides access to instances of the `Reko.Core.Program` class, which is the central clearing house for all the information Reko gathers during decompilation. Here you will find the `SegmentMap` which contails all memory segments, the `Procedures` collection that contains all the procedures Reko has discovered, and the `TypeStore` which contains all the data types Reko is aware of.

There is a host of classes that a GUI needs to be aware of in order to render things meningfully. These classes are all located in `Reko.Core` and include:
* `Procedure` - represents a decompiled procedure
* `Expression` - represents a decompiled expression. Some special subclasses are:
- `Constant` - support for integers, floats, strings
- `Identifier` - variable names
- `Address` - a pointer-valued constant.
* `Instruction` - represents a decompiled instruction, like assignments, call, jumps, and branches.
* `DataType` - represents data types of objects.
* `SegmentMap` - manages the list of `ImageSegment`s that constitute a decompiled program.
* `ImageMap` - a more detailed dictionary that maps address ranges to either a data type or a `Basic Block`
There are lots more in the Reko project, but these are the central ones around which a GUI would revolve. If you need to build a MVVM view model, these are the objects that would be rendered in it.

A hosting environment (current the `CmdLine` and `WindowsFormsDecompiler` drivers) is responsible for implementing these interfaces (except IDecompilerService(!)). Currently, @smx-smx and I are in the process of doing so at https://github.com/uxmal/reko-chromely. However, I would be very interested in seeing what a Blazor equivalent of this would look like. 

If you feel comfortable with it, you could either create a `blazor` branch directly in this repo, or create a separate prototype repository in github or wherever, and start working with hosting Reko inside of Blazor. Should you do this, you'll probably want to have a lot of questions answered, and I'd be glad to provide answers. If you choose making a separate repository, you can use the Reko runtime as a NuGet package in your Blazor project. I just released the first Reko NuGet today (https://www.nuget.org/packages/Reko.Decompiler.Runtime).

Given the complexity of creating a new GUI, I think it's more efficient to collaborate communicate using voice -- if you're comfortable speaking technical English. If this works for you it would greatly facilitate the process. @smx-smx and I have been collaborating extensively using Skype, which is my preferred voice channel.

tl;dr - sounds awesome, let's chat!

**Comment by maximilien-noal** [comment]
Hello,

Thank you for this detailled answer.

I'm definitely interested in exploring this with Blazor. MVVM is not baked in with Blazor, and I doubt that it's a must have with it. WPF was geared towards it, but using the nuget packages or the interfaces already provides separation between the core and the GUI. On top of that, MVVM feels like more busywork to handle notifications and state change when Blazor tries to manage it on its own.

Experimenting with Blazor and the nuget package sounds like a good headstart. But aren't the interfaces preferred and classes located in Reko.Core ?

Anyway, let's chat indeed! I'm in France, so it's 7am right now. After work (around 7pm) I'll be on skype (noal.maximilien@gmail.com) with some cheap work headset (it works, but sometimes not).

To publish some more detailled about my experience, in the meantime, here's what I worked on lately:
- A Blazor server side app (which uses Grid.Blazor) - proprietary
- A Windows Forms (.NET 4) DOSBox frontend (AmpShell - already ten years old)
- A .NET Core 3.1 (C# 8 + Nullable Reference Types) WPF MVVM-Light project for characters sheets and character managment (just fixed the MDI emulation layer, MDI normally doesn't exist at all in WPF)

**Comment by maximilien-noal** [comment]
I stand corrected, here is a non-trivial app in Blazor Client Side with MVVM :
https://github.com/maraf/Money

Fine for me !

**Comment by maximilien-noal** [comment]
Issue resolved over chat.

Thanks a lot. :)

Some research links :
https://github.com/CloudNimble/BlazorEssentials
https://medium.com/young-coder/the-future-of-blazor-is-native-apps-9f4a510acdd5
https://www.nuget.org/packages/WebWindow (depends on Edge 2 on Windows - not installed by default)
On Windows, it uses WPF and WebView2 :
https://docs.microsoft.com/en-us/microsoft-edge/webview2/
